hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SelectProcessorSupplier.java

```diff
@@ -74,49 +99,47 @@ public void init(@Nonnull Context context) throws Exception {
     @Nonnull
     @Override
     public Collection<? extends Processor> get(int count) {
-        List<Processor> processors = new ArrayList<>(count);
-        for (int i = 0; i < count; i++) {
-            Processor processor = new ReadJdbcP<>(
-                    () -> dataConnection.getConnection(),
-                    (connection, parallelism, index) -> {
-                        PreparedStatement statement = connection.prepareStatement(query);
-                        List<Object> arguments = evalContext.getArguments();
-                        for (int j = 0; j < parameterPositions.length; j++) {
-                            statement.setObject(j + 1, arguments.get(parameterPositions[j]));
-                        }
-                        try {
-                            return statement.executeQuery();
-                        } catch (SQLException e) {
-                            statement.close();
-                            throw e;
-                        }
-                    },
-                    rs -> {
-                        int columnCount = rs.getMetaData().getColumnCount();
-                        Object[] row = new Object[columnCount];
-                        for (int j = 0; j < columnCount; j++) {
-                            Object value = rs.getObject(j + 1);
-                            row[j] = convertValue(value);
+        assert count == 1;
+
+        Processor processor = new ReadJdbcP<>(
+                () -> dataConnection.getConnection(),
+                (connection, parallelism, index) -> {
+                    PreparedStatement statement = connection.prepareStatement(query);
+                    List<Object> arguments = evalContext.getArguments();
+                    for (int j = 0; j < parameterPositions.length; j++) {
+                        statement.setObject(j + 1, arguments.get(parameterPositions[j]));
+                    }
+                    try {
+                        ResultSet rs = statement.executeQuery();
+                        ResultSetMetaData metaData = rs.getMetaData();
+
+                        valueGetters = new BiFunctionEx[metaData.getColumnCount()];
```

Comment:
By introducing local variable for this array, doing all assignments below on this variable and assigning it to `valueGetters` at the end, you will reduce your volatile memory writes to 1.

Reply:
Right, thanks!
See 891c2a74a33